File: quant1x/level1/hello2.rs
use super::sequence_id;
use super::BinaryStream;
use crate::level1::commands::*;
use crate::level1::protocol::{Request, RequestHeader, Response, ResponseHeader};
use encoding_rs::GBK;

#[derive(Debug, Clone)]
pub struct Hello2Request {
    pub header: RequestHeader,
    pub padding: Vec<u8>,
}

impl Request for Hello2Request {
    fn header(&self) -> &RequestHeader {
        &self.header
    }

    fn header_mut(&mut self) -> &mut RequestHeader {
        &mut self.header
    }

    fn serialize_payload(&mut self) -> Vec<u8> {
        self.header.pkg_len1 = (2 + self.padding.len()) as u16;
        self.header.pkg_len2 = self.header.pkg_len1;
        let mut stream = BinaryStream::new();
        stream.push_byte_array(&self.padding);
        stream.data().clone()
    }

    fn payload_string(&self) -> String {
        format!("Hello2Request {{ padding:{} }}", hex::encode(&self.padding))
    }
}

impl Hello2Request {
    pub fn new() -> Self {
        let mut header = RequestHeader::new();
        header.zip_flag = crate::level1::protocol::zlib_flag::UNCOMPRESSED;
        header.seq_id = sequence_id();
        header.packet_type = 0x01;
        header.method = LOGIN2;

        Hello2Request {
            header,
            padding: hex::decode("d5d0c9ccd6a4a8af0000008fc22540130000d500c9ccbdf0d7ea00000002")
                .unwrap_or_default(),
        }
    }
}

#[derive(Debug, Clone)]
pub struct Hello2Response {
    pub header: ResponseHeader,
    pub info: String,
}

impl Response for Hello2Response {
    fn header(&self) -> &ResponseHeader {
        &self.header
    }

    fn header_mut(&mut self) -> &mut ResponseHeader {
        &mut self.header
    }

    fn deserialize_body(&mut self, data: &[u8]) -> Result<(), crate::std::DeserializeError> {
        let offset = 58usize;
        if data.len() >= offset {
            let (cow, _, _) = GBK.decode(&data[offset..]);
            self.info = cow.into_owned();
        }
        Ok(())
    }

    fn body_string(&self) -> String {
        format!("Info: {}", self.info)
    }
}

impl Hello2Response {
    pub fn new() -> Self {
        Self {
            header: ResponseHeader::new(),
            info: String::new(),
        }
    }
}

File: quant1x/level1/hello2.py
# -*- coding: UTF-8 -*-
# Copyright (c) Quant1X <wangfengxy@sina.cn>.
# Licensed under the MIT License.

from __future__ import annotations

import struct
from quant1x.level1.protocol import (
    FLAG_UNCOMPRESSED,
    COMMAND_LOGIN2,
    sequence_id,
)

class Synchronize2Request:
    """
    第二次协议握手请求
    """
    def __init__(self):
        self.zip_flag = FLAG_UNCOMPRESSED
        self.seq_id = sequence_id()
        self.packet_type = 0x01
        self.body_wire_len = 0
        self.body_raw_len = 0
        self.method = COMMAND_LOGIN2
        self.padding = bytes.fromhex("d5d0c9ccd6a4a8af0000008fc22540130000d500c9ccbdf0d7ea00000002")

    def serialize(self) -> bytes:
        self.body_wire_len = 2 + len(self.padding)
        self.body_raw_len = self.body_wire_len
        
        header = struct.pack('<B I B H H H', self.zip_flag, self.seq_id, self.packet_type, self.body_wire_len, self.body_raw_len, self.method)
        return header + self.padding

class Synchronize2Response:
    """
    第二次协议握手响应
    """
    def __init__(self):
        self.info = ""

    def deserialize(self, data: bytes) -> None:
        offset = 58
        if len(data) >= offset:
            info_bytes = data[offset:]
            try:
                self.info = info_bytes.decode('gbk', errors='ignore').rstrip('\x00')
            except Exception:
                self.info = info_bytes.decode('utf-8', errors='ignore')
